composer/datasets/streaming/format.py

```diff
@@ -0,0 +1,227 @@
+from io import BufferedReader, BufferedWriter
+from typing import Dict, Sequence, Tuple
+
+import numpy as np
+
+
+def get_index_basename() -> str:
+    """Get the basename for a streaming dataset index.
+
+    Returns:
+        str: Basename of file.
+    """
+    return 'index.mds'
+
+
+def get_shard_basename(shard: int) -> str:
+    """Get the basename for a streaming dataset shard.
+
+    Args:
+        shard (int): Which shard.
+
+    Returns:
+        str: Basename of file.
+    """
+    return '%05d.mds' % shard
+
+
+def sample_dict_to_bytes(obj: Dict[str, bytes], keys: Sequence[str]) -> bytes:
+    """Dump a sample dict to bytes, given field names.
+
+    Args:
+        obj (dict): The sample dict to encode.
+        keys (list of str): The field names.
+
+    Returns:
+        bytes: The encoded sample bytes.
+    """
+    values = []
+    for key in keys:
+        value = obj[key]
+        values.append(value)
+    sizes = list(map(len, values))
+    sizes = np.array(sizes, np.int32)
+    return sizes.tobytes() + b''.join(values)
+
+
+def bytes_to_sample_dict(data: bytes, keys: Sequence[str]) -> Dict[str, bytes]:
+    """Load a sample dict from bytes and field names.
+
+    Args:
+        data (bytes): The encoded sample data.
+        keys (list of str): The field names.
+
+    Returns:
+        dict: The decoded sample dict.
+    """
+    num_values = len(keys)
+    sizes = np.frombuffer(data[:num_values * 4], np.int32)
+    ends = num_values * 4 + sizes.cumsum(0)
+    begins = ends - sizes
+    values = []
+    for begin, end in zip(begins, ends):
+        value = data[begin:end]
+        values.append(value)
+    return dict(zip(keys, values))
+
+
+class StreamingDatasetIndex(object):
+    """Streaming dataset index file, containing all the info about shards.
+
+    The shards are just dumb buffers with samples catted together. All the
+    offset info across the whole dataset is contained in the index file. Workers
+    read this file to calculate how much of which shards their slice is.
+
+    Each sample is a dict of str to bytes. All samples must contain the same
+    dict keys (fields). These strings are stored in the index file for efficiency.
+    """
+
+    def __init__(self, samples_per_shard: Sequence[int], bytes_per_shard: Sequence[int],
+                 bytes_per_sample: Sequence[int], fields: Sequence[str]) -> None:
+        """Initialize with sample statistics.
+
+        Args:
+            samples_per_shard (Sequence[int]): Number of samples of each shard.
+            bytes_per_shard (Sequence[int]): Size in bytes of each shard.
+            bytes_per_sample (Sequence[int]): Size in bytes of each sample across all shards.
+            fields (Sequence[str]): The names of the samples' fields in order.
+        """
+        self.samples_per_shard = np.asarray(samples_per_shard, np.int32)
```

Comment:
Curious about the choice to encode these values as 32 bit integers rather than 64 bit? This means that the maximum size of a shard (or sample) is only 2.1 GB, which seems small, especially on a 800 Gbps (100 GBps) infiniband backbone. Is 32 bit numbers the convention in the web dataset spec, or do we get to write the spec?

Reply:
(Firsts epoch only) shard size is directly related to (a) how choppy/laggy training is, (b) how lumpy the number of samples you have to randomly sample from, and (c) how likely/unlikely samples are to be seen together (impacts quality). I've been eyeballing shard sizes to <=32MB or even smaller to keep it smooth/random.

Note that we could allocate contiguous space for the entire epoch of shards up front, and load the shards into offsets into that, if needed.

Reply:
I could see a case where we cache datasets locally via a s3 appliance on the infiniband backbone (even if we don't do this, OSS users definitely could)

Reply:
Switching all int32 to int64 after seeing another cumsum overflow